core/src/block.rs

```diff
@@ -12,594 +11,871 @@
+        /// - There is a mismatch between candidate block previous block hash and actual latest block hash
+        /// - Block has committed transactions
+        /// - Block header transaction hashes don't match with computed transaction hashes
+        /// - Error during validation of individual transactions
+        pub(crate) fn validate_without_validating_topology(
+            self,
+            block_height: u64,
+            latest_block_hash: Option<HashOf<BlockPayload>>,
+            transaction_validator: &TransactionValidator,
+            wsv: WorldStateView,
+        ) -> Result<ValidBlock, (Self, eyre::Report)> {
+            // TODO: Is it ok to not validate topology filed of the header in block_sync?
+            self.into_v1()
+                .validate_without_validating_topology(
+                    block_height,
+                    latest_block_hash,
+                    transaction_validator,
+                    wsv,
+                )
+                .map_err(|(block, err)| (block.into(), err))
         }
     }
-}
 
-/// This trait represents the ability to revalidate a block. Should be
-/// implemented for both `PendingBlock` and `VersionedCommittedBlock`.
-pub trait Revalidate: Sized {
-    /// # Errors
-    /// - When the block is deemed invalid.
-    fn revalidate<const IS_GENESIS: bool>(
-        &self,
-        transaction_validator: &TransactionValidator,
-        wsv: WorldStateView,
-        latest_block: Option<HashOf<VersionedCommittedBlock>>,
-        block_height: u64,
-    ) -> Result<(), eyre::Report>;
-
-    /// Return whether or not the block contains transactions already committed.
-    fn has_committed_transactions(&self, wsv: &WorldStateView) -> bool;
-}
+    impl SignedBlock {
+        fn retain_verified_signatures(&mut self) -> Vec<&SignatureOf<BlockPayload>> {
+            let hash = self.hash();
+            self.signatures.retain_verified_by_hash(hash).collect()
+        }
+
+        fn validate(
+            self,
+            block_height: u64,
+            latest_block_hash: Option<HashOf<BlockPayload>>,
+            topology: &Topology,
+            transaction_validator: &TransactionValidator,
+            wsv: WorldStateView,
+        ) -> Result<ValidBlock, (Self, eyre::Report)> {
+            let actual_commit_topology = &self.header().commit_topology;
+            let expected_commit_topology = &topology.ordered_peers;
+            if actual_commit_topology != expected_commit_topology {
+                let msg = eyre!("Block topology incorrect. Expected: {expected_commit_topology:#?}, actual: {actual_commit_topology:#?}");
+                return Err((self, msg));
+            }
+
+            if !self.header().is_genesis()
+                && topology
+                    .filter_signatures_by_roles(&[Role::Leader], self.signatures())
+                    .is_empty()
+            {
+                return Err((self, eyre!("Block is not signed by the leader")));
+            }
```

Comment:
Shouldn't this check be part of `commit` method?

Reply:
no, every block should have leader signature. It will be rejected before commit if it doesn't have it. On `commit` it's guaranteed that the block has it, otherwise it's impossible to create `ValidBlock`

Reply:
I just think about of block sync.
Consider situation when peer is behind for some reason (it's current `view_change_index` is 1 for example) and rest of the network committed block (at `view_change_index` 10).
In such case it might be possible that `Leader at view_change_index 2 != Leader at view_change_index 10` and slow peer will reject such block.